homeassistant/components/asuswrt/router.py

```diff
@@ -250,14 +251,25 @@ async def setup(self) -> None:
             self._sw_v = f"{firmware['firmver']} (build {firmware['buildno']})"
 
         # Load tracked entities from registry
+        entity_reg = er.async_get(self.hass)
         track_entries = er.async_entries_for_config_entry(
-            er.async_get(self.hass),
-            self._entry.entry_id,
+            entity_reg, self._entry.entry_id
         )
         for entry in track_entries:
             if entry.domain == TRACKER_DOMAIN:
-                self._devices[entry.unique_id] = AsusWrtDevInfo(
-                    entry.unique_id, entry.original_name
+                device_mac = format_mac(entry.unique_id)
+                # migrate entity unique ID if wrong formatted
+                if device_mac != entry.unique_id:
+                    try:
+                        entity_reg.async_update_entity(
+                            entry.entity_id, new_unique_id=device_mac
+                        )
+                    except ValueError:
+                        # correct entity already exist
```

Comment:
Where is ValueError thrown? I can't see where it happens, and its not clear reading this how it happens.

Reply:
Value error is exception thrown by [EntityRegistry._async_update_entity](https://github.com/home-assistant/core/blob/394ccae8a1217fdfde32353814ce9ff582afcd79/homeassistant/helpers/entity_registry.py#L434). This could happen in case of rollback to previous version, that by the way should replicate existing devices. I'm not sure if this PR is really required, main difference is that library return MAC uppercase while `format_mac`  convert it to lowercase. By design this is required for `unique_id`, but may be it is not so important at this point?

Reply:
What is the reason to remove the existing device when the ValueError happens? If the entity already exists, now it will be removed?

Reply:
Not removing the correct one with `unique_id` already formatted with `format_mac`, but a second one than have `unique_id` with same mac but not formatted (so it is uppercase). If `async_update_entity` raise `ValueError`, means that in the registry already exist an entity with same mac properly formatted, so I remove this one because is duplicated.

Reply:
`If device_mac != entry.unique_id` is there ever a case where it should not be removed? That is, should it just always get removed rather than catching the value error?

Reply:
Try to explain with an example. Suppose that this change is availble from HA 2021.12.0 and that in Asus integration exist 1 device/entity in the registry with mac `AA:BB:CC:DD`

* user upgrade to `2021.12.0`, the code try to convert `unique_id` from `AA:BB:CC:DD` to `aa:bb:cc:dd` -> no exception here
* user rollback to `2021.11.5`, the integration create again new entity with `unique_id` `AA:BB:CC:DD` because not using `format_mac` -> at this point we have in the registry 2 entity, one with `unique_id: AA:BB:CC:DD` and another with `unique_id: aa:bb:cc:dd`  
* user upgrade to `2021.12.1`, the code try to convert `unique_id` from `AA:BB:CC:DD` to `aa:bb:cc:dd` -> this fail because that entity is already un the registry, so we just detete the duplicated entity that is trying to convert.

Reply:
Can this always call remove?  It seems like here we know the entry already exists because we got the existing unique Id from the registry already.

Reply:
If we always call `remove` we will loose current settings for all devices, that will be re-created based only on available devices (just as the integration was just confugured). This doesn't seems a great solution, because users wll loose enabled / disabled configuration for all devices, and devices that are currently not connected (means `not home`) will be simply deleted.

Reply:
Maybe you mean that we should check if `entity` with converted `unique_id` already exists in the registry and in case call `remove` instead `async_update`?

Reply:
OK, I think i was just confused, and understand better now.  I wasn't suggesting what you said, but I do really like that better since it seems a little more explicit about what the intent is (e.g. imagine ValueError is thrown for some other reason).  Does that seem possible?

Reply:
I'll check and try to rewrite....

Reply:
Ok, I rewrite this removing the exception, let me know if it's more clear.